src/gtest/test_blockdownload.cpp

```diff
@@ -0,0 +1,95 @@
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+
+#include "main.h"
+
+using namespace zen;
+
+void makeTemporaryMainChain(int trunk_size);
+
+// Checks IsInitialBlockDownload() conditions and lockIBDState latching to false
+TEST(initialblockdwnld, checkIBDState) {
+    // Init
+    auto originalTip = chainActive.Tip();
+    chainActive.SetTip(NULL);
+    SelectParams(CBaseChainParams::MAIN);
+    const CChainParams& chainParams = Params();
+
+    // 1.
+    // fImporting, fReindex and fReindexFast supposedly initialized as false
+    // Check each individual variable and restore its state
+    fImporting   = true; EXPECT_TRUE(IsInitialBlockDownload()); fImporting   = false;
+    fReindex     = true; EXPECT_TRUE(IsInitialBlockDownload()); fReindex     = false;
+    fReindexFast = true; EXPECT_TRUE(IsInitialBlockDownload()); fReindexFast = false;
+
+    // 2.
+    // Expecting that fImporting, fReindex and fReindexFast are false, check if:
+    // chainActive.Height() < Checkpoints::GetTotalBlocksEstimate(chainParams.Checkpoints())
+    fCheckpointsEnabled = true;
+    EXPECT_TRUE(IsInitialBlockDownload());
```

Comment:
I agree that comments are misleading, and have been improved in commit 8036ea80c617507a1a3fd033f29bfeaa13b08ff0
The same commit also change the order of the checks in `IsInitialBlockDownload()` to correctly check all the conditions.
However, I cannot set the conditions so that `IsInitialBlockDownload()==false` at the beginning of the test, because once the atomic variable `lockIDBState` has been set, the functions is designed to always return false.

Reply:
Yeah I agree that you do not want to call `IsInitialBlockDownload()` when it returns `false` before the end of the test. I was suggesting to set the conditions for `IsInitialBlockDownload()` to return `false`, without actually calling it, and then modify one condition at a time to verify that it actually impacts the result.

This might break with future code updates without us noticing, given that you cannot actually test that you set the correct conditions for IsInitialBlockDownload()` to return `false`. On the other hand, I believe this it is still stricter than the current tests.

Reply:
I agree: commit 12d22553057f4c551d53a17a1b29974149b34c00 modifies the test so that all conditions that may return `false` are set in advance, then each flag / if is test one at a time